gtsfm/averaging/rotation/shonan.py

```diff
@@ -56,23 +67,64 @@ def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
 
         for (i1, i2), i2Ri1 in i2Ri1_dict.items():
             if i2Ri1 is not None:
-                between_factors.append(
-                    BetweenFactorPose3(
-                        i2,
-                        i1,
-                        Pose3(
-                            i2Ri1,
-                            np.zeros(
-                                3,
-                            ),
-                        ),
-                        noise_model,
-                    )
-                )
+                # ignore translation during rotation averaging
+                i2Ti1 = Pose3(i2Ri1, np.zeros(3))
+                between_factors.append(BetweenFactorPose3(i2, i1, i2Ti1, noise_model))
 
         obj = ShonanAveraging3(between_factors, shonan_params)
 
         initial = obj.initializeRandomly()
         result_values, _ = obj.run(initial, self._p_min, self._p_max)
 
-        return [result_values.atRot3(i) for i in range(num_images)]
+        wRi_list_consecutive = [None] * num_connected_nodes
+        for i in range(num_connected_nodes):
+            if result_values.exists(i):
+                wRi_list_consecutive[i] = result_values.atRot3(i)
+
+        return wRi_list_consecutive
+
+    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
+        """Run the rotation averaging on a connected graph with arbitrary keys, where each key is a image/pose index.
+
+        Note: run() functions as a wrapper that re-orders keys to prepare a graph w/ N keys ordered [0,...,N-1].
+        All input nodes must belong to a single connected component, in order to obtain an absolute pose for each
+        camera in a single, global coordinate frame.
+
+        Args:
+            num_images: number of images. Since we have one pose per image, it is also the number of poses.
+            i2Ri1_dict: relative rotations for each image pair-edge as dictionary (i1, i2): i2Ri1.
+
+        Returns:
+            Global rotations for each camera pose, i.e. wRi, as a list. The number of entries in the list is
+                `num_images`. The list may contain `None` where the global rotation could not be computed (either
+                underconstrained system or ill-constrained system), or where the camera pose had no valid observation
+                in the input to run().
```

Comment:
in other words, no edges to or from the camera in i2Ri1_dict? Maybe that is more clear, or its just me.